vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,378 @@
+    
+    do async {
+          let mutable numberOfFails = 0
+          while not firstTimeChecked && numberOfFails < 10 do
+              try
+                  do! executeCodeLenseAsync()
+                  do! Async.Sleep(1000)
+              with
+              | e -> logErrorf "Code Lens startup failed with: %A" e
+                     numberOfFails <- numberOfFails + 1
+       } |> Async.Start
+    
+    /// Creates the code lens ui elements for the specified line
+    /// TODO, add caching to the UI elements
+    let createCodeLensUIElementByLine (line: ITextViewLine) =
+        let text = line.Extent.GetText()
+        asyncMaybe {
+            let! view = view
+            // Get the real offset so that the code lens are placed correctly
+            let offset = 
+                [0..line.Length - 1] |> Seq.tryFind (fun i -> not (Char.IsWhiteSpace (line.Start.Add(i).GetChar())))
+                |> Option.defaultValue 0
+
+            let realStart = line.Start.Add(offset)
+            let span = SnapshotSpan(line.Snapshot, Span.FromBounds(int realStart, int line.End))
+            let geometry = view.TextViewLines.GetMarkerGeometry(span)
+            let! lens = codeLensResults.TryFind(text) // Check whether this line has code lens
+            if lens.Computed && codeLensUIElementCache.ContainsKey(text) then
+                // Use existing UI element which is proved to be safe to use
+                let textBox = codeLensUIElementCache.[text]
+                Canvas.SetLeft(textBox, geometry.Bounds.Left)
+                Canvas.SetTop(textBox, geometry.Bounds.Top - 15.)
+                return Some textBox
+            else if not lens.Computed then
+                lens.TaggedText |> Async.Ignore |> RoslynHelpers.StartAsyncSafe CancellationToken.None
+                let unusedElement = null
+                if codeLensUIElementCache.TryGetValue(text, &unusedElement) && not(isNull(unusedElement)) then
+                    return Some unusedElement
+                else
+                    return None
+            else
+                // No delay because it's already computed
+                let! taggedText, navigation = lens.TaggedText
+                
+                let textBox = new TextBlock(Width = view.ViewportWidth, Background = Brushes.Transparent, Opacity = 0.5, TextTrimming = TextTrimming.WordEllipsis)
+                DependencyObjectExtensions.SetDefaultTextProperties(textBox, formatMap.Value)
+                
+                for text in taggedText do
+                    let run = Documents.Run text.Text
+                    DependencyObjectExtensions.SetTextProperties (run, layoutTagToFormatting text.Tag)
+                    let inl =
+                        match text with
+                        | :? Layout.NavigableTaggedText as nav when navigation.IsTargetValid nav.Range ->                        
+                            let h = Documents.Hyperlink(run, ToolTip = nav.Range.FileName)
```

Comment:
this is where we can add the proper tooltip and the mouseOver logic to show the underline, possibly using `QuickInfoViewProvider` ?